somatic_llr_filter.py

```diff
@@ -0,0 +1,303 @@
+
+        max_llr_def = (max_llr is not None)
+        max2_llr_def = (max2_llr is not None)
+
+        if not (max_llr_def and (llr < max_llr)):
+            #save the second-place, store the new best
+            max2_call = max_call
+            max2_llr = max_llr
+            max_llr = llr
+            max_call = call
+        elif not (max2_llr_def and (llr < max2_llr)):
+            max2_call = call
+            max2_llr = llr
+    #end for
+
+    #get the final llr
+    if max_llr is not None:
+        llr = max_llr-max2_llr
+    else:
+        llr = 0
+
+    #get the final call
+    if max_call is None:
+        max_call = "-"
+
+    return [llr,max_call]
+
+
+
+def main(args_input = sys.argv[1:]):
+    parser = define_parser()
+    args = parser.parse_args(args_input)
+
+    #these aren't going to change per site
+    homozygous_expect = 1 - args.sequence_error_rate;
+    heterozygous_expect = 0.5;
+
+    vcf_reader  = create_vcf_reader(args)
+    vcf_writer = create_vcf_writer(args, vcf_reader)
+
+    for entry in vcf_reader:
+        #collect the needed info
+        ref = entry.REF
+        alts = entry.ALT
+        print(alts)
+
+        #this code will mostly handle multiple alleles, as written, but the issue is that there is
+        #no way to set the INFO (SOMATIC) field appropriately when there are multiple alleles per line,
+        # so we're going to restrict it to decomposed VCFs - one variant per line
+        if len(alts) > 1:
+             raise Exception("site with multiple alleles detected. This tool requires a decomposed vcf (one allele per line) so that INFO fields can be set appropriately")
+
+        ad_nrm = entry.call_for_sample[args.normal_sample_name].data[args.allele_depth_field]
+        ad_tum = entry.call_for_sample[args.tumor_sample_name].data[args.allele_depth_field]
+        normal_depth = entry.call_for_sample[args.tumor_sample_name].data[args.site_depth_field]
+        tumor_depth = entry.call_for_sample[args.tumor_sample_name].data[args.site_depth_field]
+        normal_ref = ad_nrm[0]
+        tumor_ref = ad_tum[0]
+
+        #TODO parse out per alt, retrieve calls
+        call = []
+        for i in range(1,(len(alts)+1)):
+            normal_var = ad_nrm[i]
+            tumor_var = ad_tum[i]
+
+            #weighted average of the frequencies
+            error_expect = (tumor_var + normal_var)/(tumor_depth + normal_depth)
+            #dave had this line in there, but I don't know why...
+            #error_expect ||= error_rate
+            if error_expect == 1:
+                error_expect = 1 - args.sequence_error_rate
+
+            tumor_freq = tumor_var/tumor_depth
+            if tumor_freq == 0:
+                tumor_freq = args.sequence_error_rate
+            elif tumor_freq == 1:
+                tumor_freq = tumor_freq - args.sequence_error_rate
+
+            call = make_call(normal_ref, normal_var, tumor_ref, tumor_var, args.sequence_error_rate, heterozygous_expect,
```

Comment:
To make this a little less ugly (i.e. not have to pass a ton of parameters to this method), I wonder if it would make sense to have a new class that has a VCF entry and calculates and stores on its instances all of these calculated values. Then you would need to just make an instance of that class with the current `entry` and you can pass along that new object to `make_call` and get all of these values directly from the instance inside of the `make_call` method.

Reply:
It probably would make things cleander, but unless you're gonna insist on it, I'm going to punt on it for now!